blscurve/bls_batch_verifier.nim

```diff
@@ -0,0 +1,376 @@
+# Nim-BLSCurve
+# Copyright (c) 2018-Present Status Research & Development GmbH
+# Licensed under either of
+#  * Apache License, version 2.0, ([LICENSE-APACHE](LICENSE-APACHE))
+#  * MIT license ([LICENSE-MIT](LICENSE-MIT))
+# at your option.
+# This file may not be copied, modified, or distributed except according to
+# those terms.
+
+import
+  ./bls_backend, ./bls_sig_min_pubkey,
+  ./openmp
+
+# BLS Batch Verifier
+# ----------------------------------------------------------------------
+# We use OpenMP here but might want to use a simple threadpool
+# for portability as Mac compiler doesn't ship with OpenMP
+# and while MSVC has OpenMP, it's unsure about Mingw
+#
+# Also as Nim supports for view types and openarray in object fields is experimental
+# we collect/copy the inputs in the object for now by copy.
+# instead of accumulating them in a pairing context.
+# Note that a pairing context is 3+MB
+# while we have
+# - publickey = 48B
+# - signature = 96B
+# - message = 32B (assuming sha256 hashed)
+# hence 176B only
+#
+# TODO: Once ContextMultiAggregateVerify is implemented
+# for Milagro/Miracl, this wouldn't need to be in the BLST specific file
+
+type
+  SignatureSet[HashLen: static int] = object
+    ## A (Public Key, Message, Signature) triplet
+    ## that will be batch verified.
+    ## This should not hold GC-ed memory
+    ## as this would complexify multithreading.
+    ## Consequently this assumes that message
+    ## is the output of a fixed size hash function.
+    signature: Signature
+    pubkey: PublicKey
+    message: array[HashLen, byte]
+
+  BatchedBLSVerifier*[HashLen: static int] = object
+    ## A type to batch BLS multi signatures (aggregated or individual)
+    ## verification using multiple cores if compiled with OpenMP
+    sets: seq[SignatureSet[HashLen]]
+
+    # Per-batch contexts for multithreaded batch verification
+    batchContexts: seq[ContextMultiAggregateVerify[DST]]
```

Comment:
why not a local variable in `batchverifyparallel`?

Reply:
Each ContextMultiAggregateVerify is a 3+MB (used to be 5+MB) because it can holds up to 8 Public Key and Signatures.

https://github.com/status-im/nim-blscurve/blob/31ada540789fc0081b4fcfd6f693410683235fd7/blscurve/blst/blst_abi.nim#L128-L158

For Miracl also it is huge as it holds 65 line functions (Fp12 elements, which are 65*12*384 bits = ~300kB)
https://github.com/status-im/nim-blscurve/blob/31ada540789fc0081b4fcfd6f693410683235fd7/blscurve/miracl/csources/64/pair_BLS12381.c#L158-L174
https://github.com/status-im/nim-blscurve/blob/9b1cf54f47e704ed52c70ef55674dbfbe0aefa8d/blscurve/miracl/csources/32/config_curve_BLS12381.h#L51

Allocating 3MB over and over will unnecessarily stress the memory allocator so it's better to reuse the same buffer.
Note that the benchmark can easily be modified to analyze the worst-case impact.

Reply:
so.. to use this correctly, one should keep a verifier instance around long-term, then clear it before every use? in that case, it seems easier to have a `verifiercache` instance and pass an `signatures: openArray[SingatureSet], contextCache: blscache` - remembering to clear before every call feel error prone